File: src/domain/usecases/auth/ObserveSessionUseCase.ts
import type { AuthRepository } from '@/domain/repositories/AuthRepository'
import type { UserRepository } from '@/domain/repositories/UserRepository'
import type { User } from '@/domain/entities/User'
import {
  CONFIGURED_SUPER_ADMIN_EMAIL,
  UserRole,
} from '@/domain/value-objects/UserRole'

export class ObserveSessionUseCase {
  private readonly authRepository: AuthRepository
  private readonly userRepository: UserRepository

  constructor(authRepository: AuthRepository, userRepository: UserRepository) {
    this.authRepository = authRepository
    this.userRepository = userRepository
  }

  execute(onChange: (user: User | null) => void): () => void {
    return this.authRepository.observeAuthState((userId) => {
      void (async () => {
        if (!userId) {
          onChange(null)
          return
        }

        let user = await this.userRepository.getById(userId)
        if (!user || !user.active) {
          await this.authRepository.logout()
          onChange(null)
          return
        }

        if (
          user.email === CONFIGURED_SUPER_ADMIN_EMAIL &&
          user.role !== UserRole.SuperAdministrador
        ) {
          try {
            await this.authRepository.claimConfiguredSuperAdmin()
            user =
              (await this.userRepository.getById(userId)) ?? {
                ...user,
                role: UserRole.SuperAdministrador,
              }
          } catch {
            // Si la función aún no está desplegada, el siguiente ingreso lo reintenta.
          }
        }

        onChange(user)
      })()
    })
  }
}

File: consorcio/lib/domain/usecases/observe_session_use_case.dart
import 'dart:async';

import '../entities/app_user.dart';
import '../errors/domain_exception.dart';
import '../repositories/auth_repository.dart';
import '../repositories/user_repository.dart';

class ObserveSessionUseCase {
  ObserveSessionUseCase(this._authRepository, this._userRepository);

  final AuthRepository _authRepository;
  final UserRepository _userRepository;

  /// Emite el usuario móvil activo. Errores de red se reportan con
  /// [Stream.addError] sin cerrar la escucha de Auth (se puede reintentar).
  Stream<AppUser?> execute() {
    late final StreamController<AppUser?> controller;
    StreamSubscription<String?>? authSub;

    controller = StreamController<AppUser?>(
      onListen: () {
        authSub = _authRepository.observeAuthState().listen(
          (userId) async {
            if (userId == null) {
              if (!controller.isClosed) controller.add(null);
              return;
            }

            try {
              final user = await _userRepository.getById(userId);
              if (user == null || !user.active || user.mobileRoles.isEmpty) {
                await _authRepository.logout();
                if (!controller.isClosed) controller.add(null);
                return;
              }
              if (!controller.isClosed) controller.add(user);
            } on DomainException catch (error) {
              if (!controller.isClosed) controller.addError(error);
            } catch (_) {
              if (!controller.isClosed) {
                controller.addError(
                  DomainException(
                    'Red lenta al leer tu perfil. Intenta de nuevo.',
                  ),
                );
              }
            }
          },
          onError: (Object error, StackTrace stackTrace) {
            if (!controller.isClosed) {
              controller.addError(error, stackTrace);
            }
          },
          onDone: () {
            if (!controller.isClosed) controller.close();
          },
        );
      },
      onCancel: () {
        unawaited(authSub?.cancel());
      },
    );

    return controller.stream;
  }
}
